cmd/vspd/spentticket.go

```diff
@@ -0,0 +1,164 @@
+// Use of this source code is governed by an ISC
+// license that can be found in the LICENSE file.package main
+
+package main
+
+import (
+	"encoding/hex"
+	"fmt"
+
+	"github.com/decred/dcrd/blockchain/stake/v5"
+	"github.com/decred/dcrd/rpc/jsonrpc/types/v4"
+	"github.com/decred/dcrd/txscript/v4/stdaddr"
+	"github.com/decred/dcrd/wire"
+	"github.com/decred/vspd/database"
+)
+
+type spentTicket struct {
+	dbTicket     database.Ticket
+	expiryHeight int64
+	heightSpent  int64
+	spendingTx   *wire.MsgTx
+}
+
+func (s *spentTicket) voted() bool {
+	return stake.IsSSGen(s.spendingTx)
+}
+
+// findSpentTickets attempts to find transactions that vote/revoke the provided
+// tickets by matching the payment script of the tickets commitment address
+// against the block filters of the mainchain blocks between the provided start
+// block and the current best block. Returns any found spent tickets and the
+// height of the most recent scanned block.
+func (v *vspd) findSpentTickets(toCheck database.TicketList, startHeight int64) (map[string]spentTicket, int64, error) {
+	params := v.cfg.netParams
+
+	dcrdClient, _, err := v.dcrd.Client()
+	if err != nil {
+		return nil, 0, err
+	}
+
+	bestBlock, err := dcrdClient.GetBestBlockHeader()
+	if err != nil {
+		return nil, 0, fmt.Errorf("dcrd.GetBestBlockHeader error: %w", err)
+	}
+	endHeight := int64(bestBlock.Height)
+
+	if startHeight > endHeight {
+		return nil, 0, fmt.Errorf("start height %d greater than best block height %d",
+			startHeight, endHeight)
+	}
+
+	numBlocks := int(1 + endHeight - startHeight)
+
+	// Only log if checking a larger number of blocks to avoid spam.
+	if numBlocks > 5 {
+		v.log.Debugf("Scanning %d blocks for %s",
+			numBlocks, pluralize(len(toCheck), "spent ticket"))
+	}
+
+	// Get commitment address payment script for each ticket.
+	type ticketTuple struct {
+		dbTicket database.Ticket
+		pkScript []byte
+	}
+
+	tickets := make(map[string]ticketTuple)
+	for _, ticket := range toCheck {
+		parsedAddr, err := stdaddr.DecodeAddress(ticket.CommitmentAddress, params)
+		if err != nil {
+			return nil, 0, err
+		}
+		_, script := parsedAddr.PaymentScript()
+
+		tickets[ticket.Hash] = ticketTuple{
+			dbTicket: ticket,
+			pkScript: script,
+		}
+	}
+
+	spent := make(map[string]spentTicket)
+
+	for iHeight := startHeight; iHeight <= endHeight; iHeight++ {
+		iHash, err := dcrdClient.GetBlockHash(iHeight)
+		if err != nil {
+			return nil, 0, err
+		}
+
+		iBlock, err := dcrdClient.GetBlockVerbose(iHash)
+		if err != nil {
+			return nil, 0, fmt.Errorf("error retrieving block %s: %w", iHash, err)
+		}
+
+		key, filter, err := dcrdClient.GetCFilterV2(iBlock.Hash, iBlock.MerkleRoot)
+		if err != nil {
+			return nil, 0, err
+		}
+
+	outer:
+		for ticketHash, ticket := range tickets {
+			if filter.Match(key, ticket.pkScript) {
+				// Ticket is likely spent in block. Confirm.
+				for _, blkTx := range append(iBlock.RawTx, iBlock.RawSTx...) {
```

Comment:
Tickets can only be spent by transactions in the stake tree (namely a vote or a revocation).  There isn't any need to range over the regular txns.